src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -361,17 +372,20 @@ private static void GetObjectAssignmentDiagnostics(ITypeManager typeManager, Obj
                         typeMismatchErrorFactory = (expectedType, actualType, errorExpression) => DiagnosticBuilder.ForPosition(errorExpression).ExpectedValueTypeMismatch(expectedType, actualType);
                     }
 
-                    GetExpressionAssignmentDiagnosticsInternal(
+                    var narrowedProperty = NarrowTypeInternal(
                         typeManager,
                         extraProperty.Value,
                         targetType.AdditionalPropertiesType.Type,
                         diagnostics,
                         typeMismatchErrorFactory,
                         skipConstantCheckForProperty,
                         skipTypeErrors: true);
+
+                    // TODO should we try and narrow the additional properties type? May be difficult
```

Comment:
I'm fine if that's done separately, but wouldn't that just be a union of types of values of "additional" properties? (Plus assignability check to verify that the union can be assigned to the type's declared `additionPropertiesType`?)

Reply:
Yeah, it would be straightforward to implement. The stuff I was concerned about:
1. How would a potentially massive union type be displayed to the user?
2. What would property access validation look like - .e.g `({a: b} | {b: c} | {c: d}).b`

Reply:
Yeah...